oceanspy/utils.py

```diff
@@ -586,3 +611,439 @@ def Coriolis_parameter(Y, omega=7.2921e-5):
+            * deg2m
+            / Delta_y
+        )
+
+    return rx, ry, cs, sn, dx, dy
+
+
+@compilable
+def find_rel_h_without_face(Xs, Ys, some_x, some_y, some_dx, some_dy, CS, SN, iys, ixs):
+    """
+    read find_rel_h for more info,
+
+    """
+    n = len(Xs)
+    rx = np.ones_like(Xs) * 0.0
+    ry = np.ones_like(Ys) * 0.0
+    dx = np.ones_like(Xs) * 0.0
+    dy = np.ones_like(Ys) * 0.0
+    cs = np.ones_like(Xs) * 0.0
+    sn = np.ones_like(Ys) * 0.0
+    for i in range(n):
+        base_lon = some_x[iys[i], ixs[i]]
+        base_lat = some_y[iys[i], ixs[i]]
+
+        cs[i] = CS[iys[i], ixs[i]]
+        sn[i] = SN[iys[i], ixs[i]]
+
+        Delta_x = some_dx[iys[i], ixs[i]]
+        Delta_y = some_dy[iys[i], ixs[i]]
+
+        dlon = to_180(Xs[i] - base_lon)
+        dlat = to_180(Ys[i] - base_lat)
+
+        dx[i] = Delta_x
+        dy[i] = Delta_y
+
+        rx[i] = (
+            (dlon * np.cos(base_lat * np.pi / 180) * cs[i] + dlat * sn[i])
+            * deg2m
+            / Delta_x
+        )
+        ry[i] = (
+            (dlat * cs[i] - dlon * sn[i] * np.cos(base_lat * np.pi / 180))
+            * deg2m
+            / Delta_y
+        )
+
+    return rx, ry, cs, sn, dx, dy
+
+
+def find_rel_h(Xs, Ys, some_x, some_y, some_dx, some_dy, CS, SN, tree):
+    """
+    very similar to find_rel_time/v
+    rx,ry,dx,dy are defined the same way
+    for example
+    rx = "how much to the right of the node"/"size of the cell in left-right direction"
+    dx = "size of the cell in left-right direction"
+
+    cs,sn is just the cos and sin of the grid orientation.
+    It will come in handy when we transfer vectors.
+    """
+    h_shape = some_x.shape
+    faces, iys, ixs = find_ind_h(Xs, Ys, tree, h_shape)
+    if faces is not None:
+        rx, ry, cs, sn, dx, dy = find_rel_h_with_face(
+            Xs, Ys, some_x, some_y, some_dx, some_dy, CS, SN, faces, iys, ixs
+        )
+    else:
+        rx, ry, cs, sn, dx, dy = find_rel_h_without_face(
+            Xs, Ys, some_x, some_y, some_dx, some_dy, CS, SN, iys, ixs
+        )
+    return faces, iys, ixs, rx, ry, cs, sn, dx, dy
+
+
+def find_rel_2d(Xs, Ys, od=None, gridtype="C"):
+    # give find_rel_h a new cover
+    global dXC, dYC, dZ, dXG, dYG, dZl, XC, XG, YC, YG, Z, Zl, CS, SN, tree
+    if od is not None:
+        dXC = np.array(od._ds["dxC"]).astype("float32")
+        dYC = np.array(od._ds["dyC"]).astype("float32")
+        dXG = np.array(od._ds["dxG"]).astype("float32")
+        dYG = np.array(od._ds["dyG"]).astype("float32")
+
+        XC = np.array(od._ds.XC).astype("float32")
+        YC = np.array(od._ds.YC).astype("float32")
+        XG = np.array(od._ds.XG).astype("float32")
+        YG = np.array(od._ds.YG).astype("float32")
+
+        CS = np.array(od._ds.CS).astype("float32")
+        SN = np.array(od._ds.SN).astype("float32")
+        tree = od.create_tree(gridtype)
+    if gridtype == "C":
```

Comment:
This can create confusion, since mitgcm (and oceanspy) is only compatible with c-grids. Instead of `gridtype`, use something like `gridloc` (as in location) or something. We don't want to suggest that these are different types of grids...